client/state/action-types.js

```diff
@@ -77,6 +77,9 @@ export const BILLING_TRANSACTIONS_RECEIVE = 'BILLING_TRANSACTIONS_RECEIVE';
 export const BILLING_TRANSACTIONS_REQUEST = 'BILLING_TRANSACTIONS_REQUEST';
 export const BILLING_TRANSACTIONS_REQUEST_FAILURE = 'BILLING_TRANSACTIONS_REQUEST_FAILURE';
 export const BILLING_TRANSACTIONS_REQUEST_SUCCESS = 'BILLING_TRANSACTIONS_REQUEST_SUCCESS';
+export const CONCIERGE_SHIFTS_FETCH = 'CONCIERGE_SHIFTS_FETCH';
+export const CONCIERGE_SHIFTS_FETCH_FAILED = 'CONCIERGE_SHIFTS_FETCH_FAILED';
+export const CONCIERGE_SHIFTS_FETCH_SUCCESS = 'CONCIERGE_SHIFTS_FETCH_SUCCESS';
```

Comment:
it looks like this is a work in progress, but take note that we likely won't need these `_FAILURE` and `_SUCCESS` action types - `dispatchRequest()` reuses the original and in many cases our reducers can _infer_ the success/failure state from normal activity, or at a minimum we tend to dispatch something like a simple notice on failure and suggest that our customers retry their operation (without needing to track error state in the app)

Reply:
@dmsnell To make sure I understand — since the `http` helper adds succes/error metadata to the action before passing it back to the handlers, we don't need the `_SUCCESS`/`_FAILURE` actions because we can simply re-dispatch `CONCIERGE_SHIFTS_FETCH` but with `meta.dataLayer.data` or `meta.dataLayer.error` auto-attached?

So we'd expect the reducers to see `CONCIERGE_SHIFTS_FETCH` twice, once for the initial fetch and once for the success or error, but they can differentiate based on the metadata in the action.

And assuming we don't need any special logic, we can `bypassDataLayer` on the action we pass to the `http` helper so that the success/error dispatch doesn't even come back through the data layer, right?

```js
import { bypassDataLayer } from 'state/data-layer/utils';
import { http } from 'state/data-layer/wpcom-http/actions';
import { dispatchRequest } from 'state/data-layer/wpcom-http/utils';
import { CONCIERGE_SHIFTS_FETCH } from 'state/action-types';

export const handleFetch = ( { dispatch }, action ) => {
	// ...
	dispatch(
		http(
			{ /* request params ... */ },
			// Bypass the data layer on the action we pass in, so that when
			// the success/error metadata is attached and it's re-dispatched
			// it won't be processed here at all
			bypassDataLayer( action )
		)
	);
};

export default {
	[ CONCIERGE_SHIFTS_FETCH ]: [
		// We don't need to pass success/error handlers to dispatchRequest
		// because the success/error action will bypass the data layer
		dispatchRequest( handleFetch )
	],
};
```

I'm having a hard time grokking everything since I haven't worked with the data layer in a while and it looks like there's been changes in the past week that make the docs outdated. Please let me know if this is the idea you were thinking about, or if I'm way off :)

Reply:
Hopefully it's simpler than it seems here. Ideally we have separate semantic action types for _requesting_ data from the server and _actually adding in data to the store_. Think of it as the difference between asking for a list of resumes and posting resumes to the hiring queue.

In this case we might expect a kind of `CONCIERGE_SHIFTS_REQUEST` to indicate to Calypso that we need the list of shifts. The reducers will likely ignore this action. It's mostly meaningless to the app itself. What matters instead to the app is stuff like, `CONCIERGE_SHIFT_ADD` or `CONCIERGE_SHIFT_REMOVE` or `CONCIERGE_SHIFT_UPDATE` - things that we might be doing regardless of any network activity. For example, we might run this from the console dispatcher in order to test a specific app state while working on the UI.

So the data layer with `dispatchRequest()` can reuse the `_REQUEST` action for the lifecycle of the network request. It may not go out immediately or at all (which is what differentiates a _request_ from a _fetch_). When it comes back in we instruct the data layer how to translate that API data into meaningful actions for Calypso.

Thus we _will_ have three functions at play here: a function to initiate network activity (by describing a network request), a function to generate Redux actions when the network activity is successful, and a function to generate actions when it fails. The data layer will do its own work on the network bits to try and keep things moving smoothly across temporary network failures and it will prevent sending out the same request at the same time.

```js
dispatchRequest( fetchShifts, storeShifts, announceFailure )
```

^^^ that's a typical pattern. `fetchShifts()` dispatches an `http()` action (which needs no bypass - the _request_ action itself should be handled here on return), the `storeShifts()` translates data from the API into data that makes sense for Calypso and dispatches those actions, then the announcer typically just sets up an `errorNotice()` indicating things failed. It's nice to alert our customers of the failure, and in most cases it's easiest just to try again so that we don't try to track complicated failure state.

The [countries list](https://github.com/Automattic/wp-calypso/blob/ff03397bbca59c5d56cfbc494123d9ddc7fa7a7d/client/state/data-layer/wpcom/domains/countries-list/index.js) is a simple example.

Thanks for poking me to ask about this. I'll be happy to answer any questions you may have.

Reply:
@dmsnell 👍 Awesome, that's a very clear explanation. Thanks for taking the time.

@southp In light of that, this PR could probably use some updates, so that we only have the one action representing the network request, and then probably just one more that takes the fetched data and passes it to reducers. (I'm not sure from a UI perspective how we want to handle errors yet, so we may just want to stub out an error handler that doesn't do anything at this point).